src/HOC/withSession.ts

```diff
@@ -0,0 +1,53 @@
+import {
+  type GetServerSidePropsContext,
+  type GetServerSidePropsResult,
+} from "next";
+import { getSession } from "next-auth/react";
+
+const redirectUrl = {
+  signin: "/signin",
+};
+
+type GetServerSidePropsResultType = Record<string, unknown>;
+type GetServerSidePropsFuncType =
+  GetServerSidePropsResult<GetServerSidePropsResultType>;
+
+function withSession(
+  getServerSidePropsFunc?: (
+    context: GetServerSidePropsContext
+  ) => Promise<GetServerSidePropsFuncType>
+) {
+  return async (
+    context: GetServerSidePropsContext
+  ): Promise<GetServerSidePropsFuncType> => {
+    const { req } = context;
+    const session = await getSession({ req });
+
+    if (!session && req.url !== redirectUrl.signin) {
+      return {
+        redirect: { destination: redirectUrl.signin, permanent: false },
+      };
+    }
+
+    let propsResult: GetServerSidePropsFuncType = {
+      props: {},
+    };
+
+    if (getServerSidePropsFunc) {
+      propsResult = await getServerSidePropsFunc(context);
```

Comment:
```suggestion
         const result = await getServerSidePropsFunc(context);
          Object.assign(propsResult, result);
```

Reply:
You'll be able to try this solution. If will not suit we could mutate propsResult